src/tableFormatter.ts

```diff
@@ -39,69 +40,85 @@ class MarkdownDocumentFormatter implements DocumentFormattingEditProvider {
-        content[1] = content[1].map(cell => {
-            if (/:-+:/.test(cell)) {
-                return ':---:';
-            } else if (/:-+/.test(cell)) {
-                return ':---';
-            } else if (/-+:/.test(cell)) {
-                return '---:';
-            } else if (/-+/.test(cell)) {
-                return '---';
-            }
-        });
-        let colWidth = Array(content[0].length).fill(3);
+    /**
+     * Get the normalized indentation of a table as multiple of the configured `tabSize`.
+     * The indentation is only read from the first line and normalized to the closest tab stop.
+     */
+    private getNormalizedIndentation(text: string, options: FormattingOptions) {
+        let indentRegex = new RegExp(/^(\s*)\S/u)
+        let match = text.match(indentRegex)
+        let indent = match[1].length
+        let tabStops = Math.round(indent / options.tabSize)
+
+        return " ".repeat(options.tabSize * tabStops)
+    }
+
+    private formatTable(text: string, doc: TextDocument, options: FormattingOptions) {
+        let indentation = this.getNormalizedIndentation(text, options)
+
+        let rows = []
+        let rowsNoIndentPattern = new RegExp(/^\s*(\S.*)$/gum)
+        let match = null
+        while ((match = rowsNoIndentPattern.exec(text)) !== null) {
+            rows.push(match[1])
+        }
+        // Get all cell contents, also for ugly format (no trailing |)
+        let fieldRegExp = new RegExp(/(?:\|?((?:\\\||\\\\\||`.*?`|[^\|])*)\|)|(?:\|?((?:\\\||\\\\\||`.*?`|[^\|])+))/gu)
+
+        let colWidth = []
         let cn = /[\u3000-\u9fff\uff01-\uff60‘“’”—]/g;
-        content.forEach(row => {
-            row.forEach((cell, i) => {
-                // Treat Chinese characters as 2 English characters
-                let cellLength = cell.length;
-                if (cn.test(cell)) {
-                    cellLength += cell.match(cn).length;
-                }
-                if (colWidth[i] < cellLength) {
-                    colWidth[i] = cellLength;
+
+        let lines = rows.map(row => {
+            let field = null
+            let values = []
+            let i = 0
+            while ((field = fieldRegExp.exec(row)) !== null) {
+                let cell = null
+                if (field[1] === undefined) {
+                    if (field[2] !== undefined && field[2].trim().length > 0) {
+                        cell = field[2].trim()
+                    } else {
+                        continue;
+                    }
+                } else {
+                    cell = field[1].trim()
                 }
-            });
+                values.push(cell)
+                // Treat Chinese characters as 2 English ones because of Unicode stuff
+                let length = cn.test(cell) ? cell.length + cell.match(cn).length : cell.length
+                colWidth[i] = colWidth[i] > length ? colWidth[i] : length
+
+                i = i + 1
+            }
```

Comment:
I am confused with this loop. The `field` variable seems to have strange semantics...

Reply:
I agree, this is confusing. The regex pattern is `(A)|(B)` hitting B only in case of an open end table (`fieldA | fieldB`).
In that case the 2nd group matches (`field[2]`) instead of 1st group (`field[1]`)